src/economy.cpp

```diff
@@ -103,38 +103,61 @@ static PriceMultipliers _price_base_multiplier;
 
 /**
  * Calculate the value of the company. That is the value of all
- * assets (vehicles, stations, etc) and money minus the loan,
+ * assets (vehicles, stations, shares) and money minus the loan,
  * except when including_loan is \c false which is useful when
  * we want to calculate the value for bankruptcy.
- * @param c              the company to get the value of.
+ * @param c the company to get the value of.
  * @param including_loan include the loan in the company value.
  * @return the value of the company.
  */
 Money CalculateCompanyValue(const Company *c, bool including_loan)
+{
+	Money companyValue = 0;
+
+	Money companyValuesExcludingShares[MAX_COMPANIES];
+
+	for (const Company* co : Company::Iterate()) {
+	  companyValuesExcludingShares[co->index]	= CalculateCompanyValueExcludingShares(co);
+	}
+
+	for (const Company* co : Company::Iterate()) {
+		for (int i = 0; i < 4; i++){
+			if (co->share_owners[i] == c->index) {
+				companyValue += (companyValuesExcludingShares[co->index] / 4);
```

Comment:
Why calculate and store company values that aren't even going to be used? CalculateCompanyValueExcludingShares can just be called here once for each relevant company.

Reply:
I wasn't sure how to get which shares the relevant company owns without iterating everything first. On this particular line though (126), CalculateCompanyValueExcludingShares will be called up to 3 times for each 25% share, so it would be  called multiple times for the same company, no ?

Reply:
they're all the same, you can just count and multiply

Reply:
OK, I committed with the implementation I think you are recommending.